From a5ec1dbef46b6e176399b85c6d24552915cf3343 Mon Sep 17 00:00:00 2001
From: Zhang Rui <bbcallen@gmail.com>
Date: Tue, 7 Jun 2016 15:38:26 +0800
Subject: [PATCH 01/14] ios/IJKFFMoviePlayerController: always calculate
 prepare delay

---
 .../IJKMediaPlayer/IJKFFMoviePlayerController.m           | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

diff --git a/ios/IJKMediaPlayer/IJKMediaPlayer/IJKFFMoviePlayerController.m b/ios/IJKMediaPlayer/IJKMediaPlayer/IJKFFMoviePlayerController.m
index bd317dc..4eb8063 100644
--- a/ios/IJKMediaPlayer/IJKMediaPlayer/IJKFFMoviePlayerController.m
+++ b/ios/IJKMediaPlayer/IJKMediaPlayer/IJKFFMoviePlayerController.m
@@ -1198,6 +1198,7 @@ static int onInjectOnHttpEvent(IJKFFMoviePlayerController *mpc, int type, void *
     IJKFFMonitor *monitor = mpc->_monitor;
     NSString     *url  = monitor.httpUrl;
     NSString     *host = monitor.httpHost;
+    int64_t       elapsed = 0;
 
     id<IJKMediaNativeInvokeDelegate> delegate = mpc.nativeInvokeDelegate;
 
@@ -1218,16 +1219,15 @@ static int onInjectOnHttpEvent(IJKFFMoviePlayerController *mpc, int type, void *
             }
             break;
         case IJKAVINJECT_DID_HTTP_OPEN:
+            elapsed = calculateElapsed(monitor.httpOpenTick, SDL_GetTickHR());
             monitor.httpError = realData->error;
             monitor.httpCode  = realData->http_code;
             monitor.httpOpenCount++;
+            monitor.httpOpenTick = 0;
+            monitor.lastHttpOpenDuration = elapsed;
             [mpc->_glView setHudValue:@(realData->http_code).stringValue forKey:@"http"];
 
             if (delegate != nil) {
-                int64_t elapsed = calculateElapsed(monitor.httpOpenTick, SDL_GetTickHR());
-                monitor.httpOpenTick = 0;
-                monitor.lastHttpOpenDuration = elapsed;
-
                 dict[IJKMediaEventAttrKey_time_of_event]    = @(elapsed).stringValue;
                 dict[IJKMediaEventAttrKey_url]              = [NSString ijk_stringBeEmptyIfNil:monitor.httpUrl];
                 dict[IJKMediaEventAttrKey_host]             = [NSString ijk_stringBeEmptyIfNil:host];
@@ -1246,16 +1246,15 @@ static int onInjectOnHttpEvent(IJKFFMoviePlayerController *mpc, int type, void *
             }
             break;
         case IJKAVINJECT_DID_HTTP_SEEK:
+            elapsed = calculateElapsed(monitor.httpSeekTick, SDL_GetTickHR());
             monitor.httpError = realData->error;
             monitor.httpCode  = realData->http_code;
             monitor.httpSeekCount++;
+            monitor.httpSeekTick = 0;
+            monitor.lastHttpSeekDuration = elapsed;
             [mpc->_glView setHudValue:@(realData->http_code).stringValue forKey:@"http"];
 
             if (delegate != nil) {
-                int64_t elapsed = calculateElapsed(monitor.httpSeekTick, SDL_GetTickHR());
-                monitor.httpSeekTick = 0;
-                monitor.lastHttpSeekDuration = elapsed;
-
                 dict[IJKMediaEventAttrKey_time_of_event]    = @(elapsed).stringValue;
                 dict[IJKMediaEventAttrKey_url]              = [NSString ijk_stringBeEmptyIfNil:monitor.httpUrl];
                 dict[IJKMediaEventAttrKey_host]             = [NSString ijk_stringBeEmptyIfNil:host];
-- 
1.9.1

