From 69149b04d0e864a8ba6d4d52c8b1133c5f88cc4b Mon Sep 17 00:00:00 2001
From: Zhang Rui <bbcallen@gmail.com>
Date: Tue, 7 Jun 2016 15:38:56 +0800
Subject: [PATCH 02/14] ijkavformat/ijklivehook: pass serveral options down to
 internal protocol

---
 ijkmedia/ijkplayer/ijkavformat/ijklivehook.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/ijkmedia/ijkplayer/ijkavformat/ijklivehook.c b/ijkmedia/ijkplayer/ijkavformat/ijklivehook.c
index 513de26..562767e 100644
--- a/ijkmedia/ijkplayer/ijkavformat/ijklivehook.c
+++ b/ijkmedia/ijkplayer/ijkavformat/ijklivehook.c
@@ -130,6 +130,12 @@ static int open_inner(AVFormatContext *avf)
     if (c->open_opts)
         av_dict_copy(&tmp_opts, c->open_opts, 0);
 
+    av_dict_set_int(&tmp_opts, "probesize",         avf->probesize, 0);
+    av_dict_set_int(&tmp_opts, "formatprobesize",   avf->format_probesize, 0);
+    av_dict_set_int(&tmp_opts, "analyzeduration",   avf->max_analyze_duration, 0);
+    av_dict_set_int(&tmp_opts, "fpsprobesize",      avf->fps_probe_size, 0);
+    av_dict_set_int(&tmp_opts, "max_ts_probe",      avf->max_ts_probe, 0);
+
     new_avf->interrupt_callback = avf->interrupt_callback;
     ret = avformat_open_input(&new_avf, c->inject_data.url, NULL, &tmp_opts);
     if (ret < 0)
-- 
1.9.1

