From 7db028c2aff5fbbf9283ced8e8f81113bb8ba5fe Mon Sep 17 00:00:00 2001
From: guowei <guowzou@gmail.com>
Date: Mon, 13 Jun 2016 20:39:27 +0800
Subject: [PATCH 09/14] fix cache display

Signed-off-by: Zhang Rui <bbcallen@gmail.com>
---
 ijkmedia/ijkplayer/ff_ffplay.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ijkmedia/ijkplayer/ff_ffplay.c b/ijkmedia/ijkplayer/ff_ffplay.c
index 43c5408..5f11e80 100644
--- a/ijkmedia/ijkplayer/ff_ffplay.c
+++ b/ijkmedia/ijkplayer/ff_ffplay.c
@@ -3853,7 +3853,7 @@ void ffp_check_buffering_l(FFPlayer *ffp)
             av_log(ffp, AV_LOG_DEBUG, "video cache=%%%d milli:(%d/%d) bytes:(%d/%d) packet:(%d/%d)\n", video_cached_percent,
                   (int)video_cached_duration, hwm_in_ms,
                   is->videoq.size, hwm_in_bytes,
-                  is->audioq.nb_packets, MIN_FRAMES);
+                  is->videoq.nb_packets, MIN_FRAMES);
 #endif
         }
 
-- 
1.9.1

